src/orderbook.ts

```diff
@@ -99,18 +139,93 @@ export class Orderbook {
       this.addAsk(ask);
     });
   }
+
+  /**
+   * Computes the transitive closure of this orderbook (e.g. ETH/DAI) with another one (e.g. DAI/USDC).
+   * Throws if the orderbooks cannot be combined (baseToken is not equal to quoteToken)
+   * @param orderbook The orderbook for which the transitive closure will be computed
+   * @returns A new instance of an orderbook representing the resulting closure.
+   */
+  transitiveClosure(orderbook: Orderbook) {
+    if (orderbook.baseToken != this.quoteToken) {
+      throw new Error(
+        `Cannot compute transitive closure of ${this.pair()} orderbook and ${orderbook.pair()} orderbook`
+      );
+    }
+
+    const ask_closure = this.transitiveAskClosure(orderbook);
+
+    // Since bids are the asks of the inverted orderbook, computing transitive closure of bids is equivalent to
+    // 1) inverting both orderbooks
+    // 2) computing the transitive ask closure on the inverses
+    // 3) re-inverting the result
+    const bid_closure = orderbook
+      .inverted()
+      .transitiveAskClosure(this.inverted())
+      .inverted();
+
+    ask_closure.add(bid_closure);
+    return ask_closure;
+  }
+
+  private transitiveAskClosure(orderbook: Orderbook) {
+    const result = new Orderbook(this.baseToken, orderbook.quoteToken);
+
+    // Create a copy here so original orders stay untouched
+    const left_asks = Array.from(this.asks.values());
+    const right_asks = Array.from(orderbook.asks.values());
+
+    left_asks.sort(sortOffersAscending);
+    right_asks.sort(sortOffersAscending);
+
+    const left_iterator = left_asks.values();
+    const right_iterator = right_asks.values();
+
+    let right_next = right_iterator.next();
+    let left_next = left_iterator.next();
+    while (!(left_next.done || right_next.done)) {
+      const right_offer = right_next.value;
+      const left_offer = left_next.value;
+      const price = left_offer.price.mul(right_offer.price);
+      let volume;
+      const right_offer_volume_in_left_offer_base_token = right_offer.volume.div(
+        left_offer.price
+      );
```

Comment:
~~Love the `;` but @bh2smith might have a heart-attack. Not sure what `tsc` has to say though.~~

**edit**: should be the subject of a separate PR where linting and code formatting get added.

Reply:
Agree. We need to change some linter setting. Right now when I omit the `;` it gets auto-added on save.